{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "\n",
    "Date : October 21, 2018\n",
    "\n",
    "Author : Heather Landry Drexler\n",
    "\n",
    "This script will develop datasets used for the confusion matrices in Supplementary Figure 2.\n",
    "                                            \n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import pysam\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import re\n",
    "% matplotlib inline\n",
    "\n",
    "import math\n",
    "\n",
    "import pybedtools\n",
    "from pybedtools import BedTool\n",
    "\n",
    "import seaborn as sns\n",
    "\n",
    "import time\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def get_sequence_bedFile(fasta, iBAM): \n",
    "    # create a bed file with read sequence\n",
    "    new_bed = []\n",
    "\n",
    "    CigarNumtoOp = {0 : 'M',\n",
    "                    1 : 'I',\n",
    "                    2 : 'D',\n",
    "                    3 : 'N',\n",
    "                    4 : 'S',             \n",
    "                    5 : 'H',\n",
    "                    6 : 'P', \n",
    "                    7 : '=',\n",
    "                    8 : 'X'}\n",
    "\n",
    "    # read in Bam file line by line\n",
    "    for read in iBAM:\n",
    "\n",
    "        # prepare a list for splice calls\n",
    "        seq = []\n",
    "\n",
    "        # set variables for parsing the cigar string\n",
    "        pattern = re.compile('([MIDNSHPX=])')\n",
    "        Consumes_Query = [\"M\", \"I\", \"S\", \"=\", \"X\"]\n",
    "        Consumes_Reference = [\"M\", \"D\", \"N\", \"=\", \"X\"]    \n",
    "\n",
    "        cigar = read.cigar                 # get cigar string\n",
    "        chrom = read.reference_name        # get chromosome name\n",
    "        ref_loc = int(read.pos)            # get starting position of read\n",
    "        cigar_loc = 0                      # get starting position of cigar (for parsing sequence)\n",
    "        sequence = read.query_sequence     # get alignment sequence\n",
    "\n",
    "        # get read strand\n",
    "        if read.is_reverse:\n",
    "            strand = \"-\"\n",
    "        else:\n",
    "            strand = \"+\"\n",
    "\n",
    "        # loop through cigar string to get coordinates for all positions that\n",
    "        # match the cigar reference sequence\n",
    "        for cigar_Entry in read.cigar:\n",
    "            cigarOp = CigarNumtoOp[cigar_Entry[0]]\n",
    "            op_Length = cigar_Entry[1]\n",
    "            cigarOp_start=ref_loc # get the starting reference coordinate of the cigar operator\n",
    "\n",
    "            if (cigarOp in Consumes_Reference):\n",
    "                ref_loc = ref_loc + op_Length # add the cigar operator length to the current location coordinate \n",
    "\n",
    "            if (cigarOp in Consumes_Query):\n",
    "                cigar_loc = cigar_loc + op_Length # add the cigar operator length to the current location coordinate \n",
    "                cigar_seq = sequence[cigar_loc-op_Length:cigar_loc]\n",
    "\n",
    "            if (cigarOp == \"M\"):\n",
    "                start = ref_loc-op_Length\n",
    "                end = ref_loc\n",
    "                new_bed.append([chrom,start,end,cigar_seq,cigarOp,strand])\n",
    "\n",
    "    match_bedtool = BedTool(new_bed)\n",
    "\n",
    "    seq_bed_df = pd.DataFrame(new_bed)\n",
    "    seq_bed_df.columns = [\"chrom\",\"start\",\"end\",\"cigar_seq\",\"cigar\",\"strand\"]\n",
    "\n",
    "    # add the reference sequence to the dataframe\n",
    "    ref_seq_bedtool = match_bedtool.sequence(fi=fasta)\n",
    "    ref_seq_open = open(ref_seq_bedtool.seqfn)\n",
    "\n",
    "    ref_seq = []\n",
    "    for line in ref_seq_open:\n",
    "        if line[0]!=\">\":\n",
    "            seq = line[:-1]\n",
    "            ref_seq.append(seq)\n",
    "\n",
    "    seq_bed_df['ref_seq'] = ref_seq\n",
    "\n",
    "    return seq_bed_df\n",
    "\n",
    "\n",
    "def get_score_df(seq_bed_df, sample_num):\n",
    "    score = pd.DataFrame([[0 for y in range(4)] for x in range(4)])\n",
    "    score.columns = ['A','T','C','G']\n",
    "    score.index = ['A','T','C','G']\n",
    "\n",
    "    seq_bed_df = seq_bed_df.sample(n=sample_num)\n",
    "    \n",
    "    for i in range(0,len(seq_bed_df)):\n",
    "\n",
    "        cigar = seq_bed_df.iloc[i]['cigar_seq']\n",
    "        reference = seq_bed_df.iloc[i]['ref_seq']\n",
    "\n",
    "        if (len(cigar) != len(reference)):\n",
    "            print(\"ERROR: line \"+str(i))\n",
    "\n",
    "        if (len(cigar) == len(reference)):\n",
    "            for j in range(0,len(cigar)):\n",
    "                if ((cigar[j]!='N') & (reference[j]!='N')):\n",
    "                    score[cigar[j]][reference[j]] = score[cigar[j]][reference[j]] + 1\n",
    "                \n",
    "    return score\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get fasta sequence\n",
    "hg38_fasta = pybedtools.example_filename('/path/to/Homo_sapiens.GRCh38.dna.primary_assembly.fa')\n",
    "\n",
    "# get alignment file\n",
    "nascent_iBAM=pysam.Samfile(\"/path/to/K562_4sUchr_ONT_3_hg38_minimap2_uniq_sort.bam\", 'rb')\n",
    "\n",
    "# get sequence bed file\n",
    "nascent_seq_df = get_sequence_bedFile(hg38_fasta, nascent_iBAM)\n",
    "nascent_seq_df.to_csv('/path/to/nanoCOP_K562_3_confusion_matrix_seq_df.txt', sep='\\t', index=False, header=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get alignment file\n",
    "UCSCrun1_iBAM=pysam.Samfile(\"/path/to/RNA_consortium_data/UCSC_Run1_hg38_minimap2_uniq_sort.bam\", 'rb')\n",
    "\n",
    "# get sequence bed file\n",
    "UCSC_Run1_seq_df = get_sequence_bedFile(hg38_fasta, UCSCrun1_iBAM)\n",
    "UCSC_Run1_seq_df.to_csv('/path/to/UCSC_Run1_confusion_matrix_seq_df.txt', sep='\\t', index=False, header=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get alignment file\n",
    "Hopkins_Run1_iBAM=pysam.Samfile(\"/path/to/RNA_consortium_data/Hopkins_Run1_hg38_minimap2_uniq_sort.bam\", 'rb')\n",
    "\n",
    "# get sequence bed file\n",
    "Hopkins_Run1_seq_df = get_sequence_bedFile(hg38_fasta, Hopkins_Run1_iBAM)\n",
    "Hopkins_Run1_seq_df.to_csv('/path/to/Hopkins_Run1_confusion_matrix_seq_df.txt', sep='\\t', index=False, header=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
