{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "\n",
    "Date : August 1, 2019\n",
    "\n",
    "Author : Heather Landry Drexler\n",
    "\n",
    "This script will develop datasets necessary to make splicing order\n",
    "plots from nano-COP data. These files serve as input for the scripts for\n",
    "Figures 4, 5 and S6.\n",
    "                                            \n",
    "\"\"\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import pysam\n",
    "from collections import Counter\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import re\n",
    "% matplotlib inline\n",
    "\n",
    "import math\n",
    "\n",
    "import pybedtools\n",
    "from pybedtools import BedTool\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# function to get a bedtool file with splice site info from hg38 intron coordinate bed file\n",
    "def hg38_introns(intronFile):\n",
    "    # make a set for all intron coordinates\n",
    "    introns = []\n",
    "\n",
    "    # loop through a file with intron coordinates\n",
    "    # record features about the introns\n",
    "    for line in intronFile:\n",
    "        chrom = str(line.split('\\t')[0])                  # chromosome\n",
    "        start = line.split('\\t')[1]                       # start coordinate of intron (last base of exon)\n",
    "        end = line.split('\\t')[2]                         # end coordinate of intron (last base of intron)\n",
    "        name_split = line.split('\\t')[3].split('_intron') # split name file by intron\n",
    "        gene = name_split[0]                              # gene name\n",
    "        feature = name_split[1].split('_')[1]             # intron count\n",
    "        strand = line.split('\\t')[5][0]                   # strand of gene with intron\n",
    "\n",
    "        # make a key that will represent intron coordinates\n",
    "        introns.append([str(chrom),str(start),str(end),str(gene),str(feature),str(strand)])\n",
    "\n",
    "    #introns_df = pd.DataFrame(introns)\n",
    "    #introns_df.columns = [\"chrom\",\"start\",\"end\",\"gene\",\"feature\",\"strand\"]\n",
    "    introns_bedtool = BedTool(introns)\n",
    "    intronFile.close()\n",
    "    return introns_bedtool\n",
    "\n",
    "\n",
    "# function to get a bedtool file with splice site info from hg38 intron coordinate bed file\n",
    "def dmel_introns(intronFile):\n",
    "    # make a set for all intron coordinates\n",
    "    introns = []\n",
    "\n",
    "    # loop through a file with intron coordinates\n",
    "    # record features about the introns\n",
    "    for line in intronFile:\n",
    "        chrom = 'chr'+str(line.split('\\t')[0])                  # chromosome\n",
    "        start = line.split('\\t')[1]                       # start coordinate of intron (last base of exon)\n",
    "        end = line.split('\\t')[2]                         # end coordinate of intron (last base of intron)\n",
    "        name_split = line.split('\\t')[3].split('_intron') # split name file by intron\n",
    "        gene = name_split[0]                              # gene name\n",
    "        feature = name_split[1].split('_')[1]             # intron count\n",
    "        strand = line.split('\\t')[5][0]                   # strand of gene with intron\n",
    "\n",
    "        # make a key that will represent intron coordinates\n",
    "        introns.append([str(chrom),str(start),str(end),str(gene),str(feature),str(strand)])\n",
    "\n",
    "    #introns_df = pd.DataFrame(introns)\n",
    "    #introns_df.columns = [\"chrom\",\"start\",\"end\",\"gene\",\"feature\",\"strand\"]\n",
    "    introns_bedtool = BedTool(introns)\n",
    "    intronFile.close()\n",
    "    return introns_bedtool\n",
    "\n",
    "\n",
    "# function to create a dataframe with reads that span 3'SS positions\n",
    "def get_intron_intersect(introns_df, bam_file):\n",
    "    # get reads that span 3' splice sites and convert to a dataframe\n",
    "    bedFile = bam_file.bam_to_bed(cigar=True, tag='NM') # convert bam file to bed file, keep cigar string and NM (edit distance) tag\n",
    "    intersect = bedFile.intersect(introns_df, wo=True, s=True) # intersect reads from bam file with 3' splice site coordinates, ensure strandedness\n",
    "    df = intersect.to_dataframe(names=['chr_aln', 'start_aln', 'end_aln', 'name_aln', 'qual_aln', \\\n",
    "                                           'strand_aln', 'cigar_aln', 'chr_intron', 'start_intron', \\\n",
    "                                           'end_intron', 'name_gene', 'intron_count', 'strand_gene', 'count'], \\\n",
    "                               dtype={\"chr_aln\": str, \"start_aln\": int, \"end_aln\": int, \\\n",
    "                                     \"name_aln\": str, \"qual_aln\": int, \"strand_aln\": str, \\\n",
    "                                     \"cigar_aln\": str, \"chr_intron\": str, \"start_intron\": int, \\\n",
    "                                     \"end_intron\": int, \"name_gene\": str, \\\n",
    "                                     \"intron_count\": int,\"strand_gene\": str, \"count\": int}) # convert to a dataframe\n",
    "    return df\n",
    "\n",
    "\n",
    "# function to create a dataframe with splicing information for\n",
    "# every read that spans an intron in the dataset\n",
    "def get_splicing_info(intersect_df, min_overlap):\n",
    "    \n",
    "    df = intersect_df\n",
    "\n",
    "    # prepare a list for splice calls\n",
    "    spliceCalls = []\n",
    "\n",
    "    # set variables for parsing the cigar string\n",
    "    pattern = re.compile('([MIDNSHPX=])')\n",
    "    Consumes_Query = [\"M\", \"I\", \"S\", \"=\", \"X\"]\n",
    "    Consumes_Reference = [\"M\", \"D\", \"N\", \"=\", \"X\"]    \n",
    "\n",
    "    # loop through all read-intron intersects\n",
    "    for i in range(0,df.shape[0]):\n",
    "\n",
    "        # ignore reads that do not overlap intron by minimum threshold\n",
    "        if (df['count'].iloc[i] < min_overlap):\n",
    "            continue\n",
    "\n",
    "        # record the start and ends of reads \n",
    "        # will deal with gene strand after cigar counts are made\n",
    "        aln_start = df['start_aln'].iloc[i] # record the start of the read\n",
    "        aln_end = df['end_aln'].iloc[i] # record the end of the read\n",
    "        intron_start = df['start_intron'].iloc[i] # record the end of the intron\n",
    "        intron_end = df['end_intron'].iloc[i] # record the end of the intron\n",
    "\n",
    "        # parse cigar string into a list of tuples for easy parsing\n",
    "        Sep_Values = pattern.split(df['cigar_aln'].iloc[i])[:-1]\n",
    "        CigarPairs = list((Sep_Values[n:n+2] for n in range(0, len(Sep_Values), 2)))  \n",
    "\n",
    "        # set up variables for measuring the length of cigar string operators\n",
    "        CigarOp_counts = {'M': 0, 'I': 0, 'D': 0, 'N': 0, 'S': 0, 'H': 0, 'P': 0, '=': 0, 'X': 0}\n",
    "        start_counts = {'M': 0, 'I': 0, 'D': 0, 'N': 0, 'S': 0, 'H': 0, 'P': 0, '=': 0, 'X': 0}\n",
    "        end_counts = {'M': 0, 'I': 0, 'D': 0, 'N': 0, 'S': 0, 'H': 0, 'P': 0, '=': 0, 'X': 0}\n",
    "        intron_counts = {'M': 0, 'I': 0, 'D': 0, 'N': 0, 'S': 0, 'H': 0, 'P': 0, '=': 0, 'X': 0}\n",
    "        currentloc = int(df['start_aln'].iloc[i])\n",
    "\n",
    "        # go through list of cigar strings and grab splicing information\n",
    "        for cigar_Entry in CigarPairs:\n",
    "\n",
    "            op_Length = int(cigar_Entry[0]) # get length of cigar operator\n",
    "            cigarOp = cigar_Entry[1] # get type of cigar operator  \n",
    "            CigarOp_counts[cigarOp] += op_Length # add the cigar operator length to the counts dictionary\n",
    "            cigarOp_start=currentloc # get the starting coordinate of the cigar operator\n",
    "\n",
    "            if (cigarOp in Consumes_Reference):\n",
    "                currentloc=currentloc+op_Length # add the cigar operator length to the current location coordinate \n",
    "\n",
    "            cigarOp_end=currentloc # get the ending coordinate of the cigar operator\n",
    "\n",
    "            # gather information if the portion of the cigar string spans the designated intron start\n",
    "            if (cigarOp_start<intron_start-min_overlap and cigarOp_end>=intron_start-min_overlap):\n",
    "                if (cigarOp_end>=intron_start+min_overlap):\n",
    "                    count=min_overlap*2\n",
    "                else:\n",
    "                    count=cigarOp_end-(intron_start-min_overlap)+1\n",
    "                start_counts[cigarOp] += count # add the cigar operator length to the counts dictionary\n",
    "\n",
    "            elif (cigarOp_start>=intron_start-min_overlap and cigarOp_end<intron_start+min_overlap):\n",
    "                count=op_Length\n",
    "                start_counts[cigarOp] += count # add the cigar operator length to the counts dictionary       \n",
    "\n",
    "            elif (cigarOp_start<intron_start+min_overlap and cigarOp_end>=intron_start+min_overlap):\n",
    "                if (cigarOp_start<=intron_start-min_overlap):\n",
    "                    count=min_overlap*2\n",
    "                else:\n",
    "                    count=(intron_start+min_overlap)-cigarOp_start-1\n",
    "                start_counts[cigarOp] += count # add the cigar operator length to the counts dictionary \n",
    "\n",
    "            # gather information if the portion of the cigar string is within the intron\n",
    "            if (cigarOp_start<intron_start and cigarOp_end>=intron_start):\n",
    "                if (cigarOp_end>=intron_end):\n",
    "                    count=intron_end-intron_start\n",
    "                else:\n",
    "                    count=cigarOp_end-intron_start\n",
    "                intron_counts[cigarOp] += count # add the cigar operator length to the counts dictionary\n",
    "\n",
    "            elif (cigarOp_start>=intron_start and cigarOp_end<intron_end):\n",
    "                count=op_Length\n",
    "                intron_counts[cigarOp] += count # add the cigar operator length to the counts dictionary\n",
    "\n",
    "            elif (cigarOp_start<intron_end and cigarOp_end>=intron_end):\n",
    "                if (cigarOp_start<=intron_start):\n",
    "                    count=intron_end-intron_start\n",
    "                else:\n",
    "                    count=intron_end-cigarOp_start\n",
    "                intron_counts[cigarOp] += count # add the cigar operator length to the counts dictionary \n",
    "\n",
    "            # gather information if the portion of the cigar string spans the designated intron end\n",
    "            if (cigarOp_start<intron_end-min_overlap and cigarOp_end>=intron_end-min_overlap):\n",
    "                if (cigarOp_end>=intron_end+min_overlap):\n",
    "                    count=min_overlap*2\n",
    "                else:\n",
    "                    count=cigarOp_end-(intron_end-min_overlap)\n",
    "                end_counts[cigarOp] += count # add the cigar operator length to the counts dictionary\n",
    "\n",
    "            elif (cigarOp_start>=intron_end-min_overlap and cigarOp_end<intron_end+min_overlap):\n",
    "                count=op_Length\n",
    "                end_counts[cigarOp] += count # add the cigar operator length to the counts dictionary\n",
    "\n",
    "            elif (cigarOp_start<intron_end+min_overlap and cigarOp_end>=intron_end+min_overlap):\n",
    "                if (cigarOp_start<=intron_end-min_overlap):\n",
    "                    count=min_overlap*2\n",
    "                else:\n",
    "                    count=(intron_end+min_overlap)-cigarOp_start\n",
    "                end_counts[cigarOp] += count # add the cigar operator length to the counts dictionary \n",
    "\n",
    "        # get length of the aligned portion of this read from cigar string\n",
    "        aligned_read_length = CigarOp_counts['M']+CigarOp_counts['D']\n",
    "\n",
    "        # get 5'SS and 3'SS counts as determined by gene strand\n",
    "        strand = df['strand_gene'].iloc[i]\n",
    "        if (strand == '+'):\n",
    "            aln_start = df['start_aln'].iloc[i] # record the start of the read\n",
    "            aln_end = df['end_aln'].iloc[i] # record the end of the read\n",
    "            intron_5SS_counts = start_counts # record the cigar string counts over the 5'SS\n",
    "            intron_3SS_counts = end_counts # record the cigar string counts over the 3'SS\n",
    "            read_overlap = intron_end - (aln_end - aligned_read_length + min_overlap)\n",
    "            \n",
    "        if (strand == '-'):\n",
    "            aln_start = df['end_aln'].iloc[i] # record the start of the read\n",
    "            aln_end = df['start_aln'].iloc[i] # record the end of the read\n",
    "            intron_5SS_counts = end_counts # record the cigar string counts over the 5'SS\n",
    "            intron_3SS_counts = start_counts # record the cigar string counts over the 3'SS  \n",
    "            read_overlap = (aln_end + aligned_read_length - min_overlap) - intron_start\n",
    "            \n",
    "        # annotate splicing status based on CIGAR string information around splice sites\n",
    "        splice='UNDETERMINED'\n",
    "\n",
    "        if (intron_5SS_counts['N']==0 and intron_3SS_counts['N']==0):\n",
    "            if (intron_3SS_counts['M']+intron_3SS_counts['D']==min_overlap*2):\n",
    "                if (intron_3SS_counts['M']>min_overlap):\n",
    "                    splice = 'NO'\n",
    "\n",
    "        if (intron_5SS_counts['N']>0 and intron_5SS_counts['N']<min_overlap*2):\n",
    "            if (intron_3SS_counts['N']>0 and intron_3SS_counts['N']<min_overlap*2):\n",
    "                splice = 'YES'\n",
    "\n",
    "        # annotate splicing status based on CIGAR string information within the intron \n",
    "        if (splice == 'YES'):\n",
    "            if (float(intron_end-intron_start) > 0.0):\n",
    "                ratio = float(intron_counts['N'])/float(intron_end-intron_start)\n",
    "                difference = abs(intron_counts['N']-(intron_end-intron_start))\n",
    "\n",
    "                # if read is spliced, between 90-100% of the intron has to be spliced \n",
    "                # and no more than 100 nucleotides within the intron can be matching the intron sequence\n",
    "                if( ratio < 0.9 or ratio > 1.1 or difference > 100):\n",
    "                    splice='UNDETERMINED'\n",
    "            if (float(intron_end-intron_start) == 0.0):\n",
    "                splice='UNDETERMINED'\n",
    "\n",
    "        if (splice == 'NO'):\n",
    "            if (float(intron_end-intron_start) > 0.0):\n",
    "                ratio = float(intron_counts['M'])/(float(intron_counts['M'])+float(intron_counts['N'])+float(intron_counts['D'])+1)\n",
    "\n",
    "                # if read is unspliced, at least 75% of the read has to match (CIGAR=M) the intron sequence\n",
    "                if(intron_counts['M'] < min_overlap/2 or ratio < 0.75):\n",
    "                    splice='UNDETERMINED'\n",
    "            \n",
    "            if (float(intron_end-intron_start) == 0.0):\n",
    "                splice='UNDETERMINED'\n",
    "\n",
    "        # save read, intron, and splicing information\n",
    "        spliceCalls.append([df['name_aln'].iloc[i],df['chr_intron'].iloc[i],intron_start,intron_end,df['strand_gene'].iloc[i],df['name_gene'].iloc[i],df['intron_count'].iloc[i],read_overlap,splice])\n",
    "\n",
    "    spliceCalls_df = pd.DataFrame(spliceCalls)\n",
    "    spliceCalls_df.columns = [\"read_name\",\"chrom\",\"intron_start\",\"intron_end\",\"strand\",\"gene_name\",\"intron_count\",\"read_overlap\",\"splice_status\"]\n",
    "\n",
    "    return spliceCalls_df\n",
    "\n",
    "\n",
    "# every read that spans an intron in the dataset\n",
    "def get_read_junctions_dictionary(splice_df):\n",
    "\n",
    "    read_junctions = {}\n",
    "\n",
    "    for i in range(0,splice_df.shape[0]):       \n",
    "\n",
    "        # define the read name\n",
    "        read_name = splice_df['read_name'].iloc[i]\n",
    "        gene_name = splice_df['gene_name'].iloc[i]\n",
    "        chrom = splice_df['chrom'].iloc[i]\n",
    "        intron_start = splice_df['intron_start'].iloc[i]\n",
    "        intron_end = splice_df['intron_end'].iloc[i]\n",
    "        intron_count = splice_df['intron_count'].iloc[i]\n",
    "        strand = splice_df['strand'].iloc[i]\n",
    "        read_overlap = splice_df['read_overlap'].iloc[i]\n",
    "        splice_status = splice_df['splice_status'].iloc[i]\n",
    "\n",
    "        # check if read name is in the dictionary, if not save it\n",
    "        if read_name not in read_junctions.keys():\n",
    "\n",
    "            # make a new dictionary for the gene and add intron info to it\n",
    "            read_junctions[read_name] = {}\n",
    "            read_junctions[read_name][gene_name] = [[chrom, intron_start, intron_end, intron_count, strand, read_overlap, splice_status]]\n",
    "\n",
    "        # check if read name is in the dictionary, if it is proceed to gene information\n",
    "        elif read_name in read_junctions.keys():\n",
    "\n",
    "            # if gene_name is not already in read dictionary, \n",
    "            # make a new dictionary for the gene and add intron info to it\n",
    "            if gene_name not in read_junctions[read_name].keys():\n",
    "                read_junctions[read_name][gene_name] = [[chrom, intron_start, intron_end, intron_count, strand, read_overlap, splice_status]]\n",
    "\n",
    "            # if gene_name is already in read dictionary, add new intron info to it\n",
    "            elif gene_name in read_junctions[read_name].keys():\n",
    "                read_junctions[read_name][gene_name].append([chrom, intron_start, intron_end, intron_count, strand, read_overlap, splice_status])\n",
    "\n",
    "    return read_junctions\n",
    "\n",
    "\n",
    "def get_intron_pairs_df(read_junctions):\n",
    "    intron_pairs = []\n",
    "\n",
    "    # loop through all reads in the dictionary\n",
    "    for read in read_junctions.keys():\n",
    "\n",
    "        # make a set for all intron pairs within a read\n",
    "        # this will avoid duplicate pairs being called due to alternative splicing\n",
    "        uniq_pairs = set()\n",
    "        uniq_splice_pattern = set()\n",
    "        \n",
    "        # loop through all genes that has introns that a read maps to\n",
    "        for gene in read_junctions[read].keys():\n",
    "\n",
    "            # only go through genes that have 2 or more introns\n",
    "            if (len(read_junctions[read][gene]) > 1 ):\n",
    "\n",
    "                # characterize the number of spliced and unspliced introns in the read\n",
    "                splice_status = [row[6] for row in read_junctions[read][gene]]\n",
    "                splice_status_join = '_'.join(splice_status)\n",
    "                status_count = Counter(splice_status)\n",
    "                 \n",
    "                # only process the file if intron pattern hasn't been seen previously\n",
    "                # for a gene that this read aligns to\n",
    "                if (splice_status_join not in uniq_splice_pattern):\n",
    "                    uniq_splice_pattern.add(splice_status_join)\n",
    "\n",
    "                    spliced_count = status_count['YES']\n",
    "                    unspliced_count = status_count['NO']\n",
    "\n",
    "                    # build a dataframe of introns in the gene that map to this read\n",
    "                    # and are capable of being sequenced if the read has no splicing\n",
    "                    read_introns_df = pd.DataFrame(read_junctions[read][gene])\n",
    "                    read_introns_df.columns = ['chrom','start','end','intron_count','strand','read_overlap','splice_status']\n",
    "                    read_introns_df = read_introns_df[read_introns_df['read_overlap'] > 0].sort_values('intron_count').reset_index(drop=True)\n",
    "\n",
    "                    # loop through introns that read maps to and find pairs\n",
    "                    prev_intron_count = -2    # counter for the start becuase no intron should have a negative count\n",
    "\n",
    "                    for i in range(len(read_introns_df)):\n",
    "                        intron_count = read_introns_df.iloc[i]['intron_count']\n",
    "                        intron_chrom = str(read_introns_df.iloc[i]['chrom'])\n",
    "                        intron_start = str(read_introns_df.iloc[i]['start'])\n",
    "                        intron_end = str(read_introns_df.iloc[i]['end'])\n",
    "                        intron_strand = read_introns_df.iloc[i]['strand']\n",
    "                        intron_splice = read_introns_df.iloc[i]['splice_status']\n",
    "                        intron_coord = intron_chrom+'_'+intron_start+'_'+intron_end\n",
    "\n",
    "                        # if intron counts are sequential (one follows the next)\n",
    "                        # it is a true intron pair (i.e. neighboring introns)\n",
    "                        if (intron_count - prev_intron_count == 1):\n",
    "                            intron_pair_coord = prev_intron_coord+'_'+intron_coord\n",
    "\n",
    "                            # record information about the read pair only if the coordinates of this\n",
    "                            # intron pair have not yet been seen\n",
    "                            if (intron_pair_coord not in uniq_pairs): \n",
    "                                uniq_pairs.add(intron_pair_coord)\n",
    "                                prev_intron_start = prev_intron_coord.split('_')[1]\n",
    "                                prev_intron_end = prev_intron_coord.split('_')[2]\n",
    "\n",
    "                                # append intron pair coordinate and splicing information to a list\n",
    "                                intron_pairs.append([read,intron_chrom,prev_intron_start,prev_intron_end,\n",
    "                                                     int(intron_start),int(intron_end),intron_strand,\n",
    "                                                    prev_intron_splice, intron_splice])\n",
    "\n",
    "                        # save information about this intron for the next pair\n",
    "                        prev_intron_count = intron_count\n",
    "                        prev_intron_coord = intron_coord\n",
    "                        prev_intron_splice = intron_splice\n",
    "\n",
    "    intron_pairs_df = pd.DataFrame(intron_pairs)\n",
    "    intron_pairs_df.columns = ['read','chrom','int1_start','int1_end','int2_start','int2_end','strand','int1_splice','int2_splice']        \n",
    "\n",
    "    return intron_pairs_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# set all variables for analysis\n",
    "min_overlap = 25"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get human data for figure 3B (Splicing pattern proportions)\n",
    "\n",
    "# install splice sites from K562 intron files\n",
    "K562_intronFile = open('/path/to/annotation_files/NCBI_RefSeq_hg38_introns_parsed.bed')\n",
    "\n",
    "# make a dataframe of intron coordinates\n",
    "K562_introns_bedtool = hg38_introns(K562_intronFile) \n",
    "\n",
    "# import alignment files\n",
    "K562_1_bamFile = pybedtools.BedTool('/path/to/K562_4sUchr_ONT_1_hg38_minimap2_uniq_sort.bam')\n",
    "K562_2_bamFile = pybedtools.BedTool('/path/to/K562_4sUchr_ONT_2_hg38_minimap2_uniq_sort.bam')\n",
    "K562_3_bamFile = pybedtools.BedTool('/path/to/K562_4sUchr_ONT_3_hg38_minimap2_uniq_sort.bam')\n",
    "K562_4_bamFile = pybedtools.BedTool('/path/to/K562_4sUchr_ONT_4_hg38_minimap2_uniq_sort.bam')\n",
    "K562_5a_bamFile = pybedtools.BedTool('/path/to/K562_4sUchr_ONT_5a_hg38_minimap2_uniq_sort.bam')\n",
    "K562_5b_bamFile = pybedtools.BedTool('/path/to/K562_4sUchr_ONT_5b_hg38_minimap2_uniq_sort.bam')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get reads that intersect introns\n",
    "K562_1_intersect = get_intron_intersect(K562_introns_bedtool, K562_1_bam_file)\n",
    "K562_2_intersect = get_intron_intersect(K562_introns_bedtool, K562_2_bam_file)\n",
    "K562_3_intersect = get_intron_intersect(K562_introns_bedtool, K562_3_bam_file)\n",
    "K562_4_intersect = get_intron_intersect(K562_introns_bedtool, K562_4_bam_file)\n",
    "K562_5a_intersect = get_intron_intersect(K562_introns_bedtool, K562_5a_bam_file)\n",
    "K562_5b_intersect = get_intron_intersect(K562_introns_bedtool, K562_5b_bam_file)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get splicing information for every read that spans an intron\n",
    "K562_1_splice_info = get_splicing_info(K562_1_intersect,min_overlap)\n",
    "K562_2_splice_info = get_splicing_info(K562_2_intersect,min_overlap)\n",
    "K562_3_splice_info = get_splicing_info(K562_3_intersect,min_overlap)\n",
    "K562_4_splice_info = get_splicing_info(K562_4_intersect,min_overlap)\n",
    "K562_5a_splice_info = get_splicing_info(K562_5a_intersect,min_overlap)\n",
    "K562_5b_splice_info = get_splicing_info(K562_5b_intersect,min_overlap)\n",
    "\n",
    "\n",
    "# remove unwanted intron information\n",
    "K562_1_splice_info = K562_1_splice_info[K562_1_splice_info['splice_status'] != \"UNDETERMINED\"].reset_index(drop=True)\n",
    "K562_1_splice_info_wiOverlap = K562_1_splice_info[K562_1_splice_info['read_overlap']>(K562_1_splice_info['intron_end']-K562_1_splice_info['intron_start'])].reset_index(drop=True)\n",
    "\n",
    "K562_2_splice_info = K562_2_splice_info[K562_2_splice_info['splice_status'] != \"UNDETERMINED\"].reset_index(drop=True)\n",
    "K562_2_splice_info_wiOverlap = K562_2_splice_info[K562_2_splice_info['read_overlap']>(K562_2_splice_info['intron_end']-K562_2_splice_info['intron_start'])].reset_index(drop=True)\n",
    "\n",
    "K562_3_splice_info = K562_3_splice_info[K562_3_splice_info['splice_status'] != \"UNDETERMINED\"].reset_index(drop=True)\n",
    "K562_3_splice_info_wiOverlap = K562_3_splice_info[K562_3_splice_info['read_overlap']>(K562_3_splice_info['intron_end']-K562_3_splice_info['intron_start'])].reset_index(drop=True)\n",
    "\n",
    "K562_4_splice_info = K562_4_splice_info[K562_4_splice_info['splice_status'] != \"UNDETERMINED\"].reset_index(drop=True)\n",
    "K562_4_splice_info_wiOverlap = K562_4_splice_info[K562_4_splice_info['read_overlap']>(K562_4_splice_info['intron_end']-K562_4_splice_info['intron_start'])].reset_index(drop=True)\n",
    "\n",
    "K562_5a_splice_info = K562_5a_splice_info[K562_5a_splice_info['splice_status'] != \"UNDETERMINED\"].reset_index(drop=True)\n",
    "K562_5a_splice_info_wiOverlap = K562_5a_splice_info[K562_5a_splice_info['read_overlap']>(K562_5a_splice_info['intron_end']-K562_5a_splice_info['intron_start'])].reset_index(drop=True)\n",
    "\n",
    "K562_5b_splice_info = K562_5b_splice_info[K562_5b_splice_info['splice_status'] != \"UNDETERMINED\"].reset_index(drop=True)\n",
    "K562_5b_splice_info_wiOverlap = K562_5b_splice_info[K562_5b_splice_info['read_overlap']>(K562_5b_splice_info['intron_end']-K562_5b_splice_info['intron_start'])].reset_index(drop=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get dictionary with all intron junctions that a read spans\n",
    "K562_1_splice_dictionary = get_read_junctions_dictionary(K562_1_splice_info_wiOverlap)\n",
    "K562_2_splice_dictionary = get_read_junctions_dictionary(K562_2_splice_info_wiOverlap)\n",
    "K562_3_splice_dictionary = get_read_junctions_dictionary(K562_3_splice_info_wiOverlap)\n",
    "K562_4_splice_dictionary = get_read_junctions_dictionary(K562_4_splice_info_wiOverlap)\n",
    "K562_5a_splice_dictionary = get_read_junctions_dictionary(K562_5a_splice_info_wiOverlap)\n",
    "K562_5b_splice_dictionary = get_read_junctions_dictionary(K562_5b_splice_info_wiOverlap)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# save splicing dictionaries to file (to use again later for plotting) (K562 cells)\n",
    "np.save('/path/to/K562_1_splicing_dictionary.npy', K562_6_splice_dictionary) \n",
    "np.save('/path/to/K562_2_splicing_dictionary.npy', K562_7_splice_dictionary) \n",
    "np.save('/path/to/K562_3_splicing_dictionary.npy', K562_12_splice_dictionary) \n",
    "np.save('/path/to/K562_4_splicing_dictionary.npy', K562_27_splice_dictionary) \n",
    "np.save('/path/to/K562_5a_splicing_dictionary.npy', Pr1_splice_dictionary) \n",
    "np.save('/path/to/K562_5b_splicing_dictionary.npy', Pr2_splice_dictionary) \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get information about intron pairs from read junctions dictionary\n",
    "K562_1_intron_pairs_df = get_intron_pairs_df(K562_1_splice_dictionary)\n",
    "K562_2_intron_pairs_df = get_intron_pairs_df(K562_2_splice_dictionary)\n",
    "K562_3_intron_pairs_df = get_intron_pairs_df(K562_3_splice_dictionary)\n",
    "K562_4_intron_pairs_df = get_intron_pairs_df(K562_4_splice_dictionary)\n",
    "K562_5a_intron_pairs_df = get_intron_pairs_df(K562_5a_splice_dictionary)\n",
    "K562_5b_intron_pairs_df = get_intron_pairs_df(K562_5b_splice_dictionary)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# save splice dataframes to file (to use again later for plotting) (K562 cells)\n",
    "K562_1_intron_pairs_df.to_csv('/path/to/K562_1_hg38_intron_pairs_df.txt', sep='\\t', index=False, header=True)\n",
    "K562_2_intron_pairs_df.to_csv('/path/to/K562_2_hg38_intron_pairs_df.txt', sep='\\t', index=False, header=True)\n",
    "K562_3_intron_pairs_df.to_csv('/path/to/K562_3_hg38_intron_pairs_df.txt', sep='\\t', index=False, header=True)\n",
    "K562_4_intron_pairs_df.to_csv('/path/to/K562_4_hg38_intron_pairs_df.txt', sep='\\t', index=False, header=True)\n",
    "K562_5a_intron_pairs_df.to_csv('/path/to/K562_5a_hg38_intron_pairs_df.txt', sep='\\t', index=False, header=True)\n",
    "K562_5b_intron_pairs_df.to_csv('/path/to/K562_5b_hg38_intron_pairs_df.txt', sep='\\t', index=False, header=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "### intron pairs for B lymphoblast (BL1184) sample\n",
    "\n",
    "# load bam files\n",
    "BL1184_1_bam_file = pybedtools.BedTool('/path/to/BL1184_4sUchr_ONT_1_hg38_minimap2_uniq_sort.bam')\n",
    "BL1184_2_bam_file = pybedtools.BedTool('/path/to/BL1184_4sUchr_ONT_2_hg38_minimap2_uniq_sort.bam')\n",
    "\n",
    "# get reads that intersect introns\n",
    "BL1184_1_intersect = get_intron_intersect(K562_introns_bedtool, BL1184_1_bam_file)\n",
    "BL1184_2_intersect = get_intron_intersect(K562_introns_bedtool, BL1184_2_bam_file)\n",
    "\n",
    "# get splicing information for every read that spans an intron\n",
    "BL1184_1_splice_info = get_splicing_info(BL1184_1_intersect,min_overlap)\n",
    "BL1184_2_splice_info = get_splicing_info(BL1184_2_intersect,min_overlap)\n",
    "\n",
    "# remove unwanted intron information\n",
    "BL1184_1_splice_info = BL1184_1_splice_info[BL1184_1_splice_info['splice_status'] != \"UNDETERMINED\"].reset_index(drop=True)\n",
    "BL1184_1_splice_info_wiOverlap = BL1184_1_splice_info[BL1184_1_splice_info['read_overlap']>(BL1184_1_splice_info['intron_end']-BL1184_1_splice_info['intron_start'])].reset_index(drop=True)\n",
    "BL1184_2_splice_info = BL1184_2_splice_info[BL1184_2_splice_info['splice_status'] != \"UNDETERMINED\"].reset_index(drop=True)\n",
    "BL1184_2_splice_info_wiOverlap = BL1184_2_splice_info[BL1184_2_splice_info['read_overlap']>(BL1184_2_splice_info['intron_end']-BL1184_2_splice_info['intron_start'])].reset_index(drop=True)\n",
    "\n",
    "# get dictionary with all intron junctions that a read spans\n",
    "BL1184_1_splice_dictionary = get_read_junctions_dictionary(BL1184_1_splice_info_wiOverlap)\n",
    "BL1184_2_splice_dictionary = get_read_junctions_dictionary(BL1184_2_splice_info_wiOverlap)\n",
    "\n",
    "# save splicing dictionary datasets\n",
    "np.save('/path/to/BL1184_1_splicing_dictionary.npy', BL1184_1_splice_dictionary) \n",
    "np.save('/path/to/BL1184_2_splicing_dictionary.npy', BL1184_2_splice_dictionary) \n",
    "\n",
    "# get information about intron pairs from read junctions dictionary\n",
    "BL1184_1_intron_pairs_df = get_intron_pairs_df(BL1184_1_splice_dictionary)\n",
    "BL1184_2_intron_pairs_df = get_intron_pairs_df(BL1184_2_splice_dictionary)\n",
    "\n",
    "# save intron_pairs datasets\n",
    "BL1184_1_intron_pairs_df.to_csv('/path/to/BL1184_1_hg38_intron_pairs_df.txt', sep='\\t', index=False, header=True)\n",
    "BL1184_2_intron_pairs_df.to_csv('/path/to/BL1184_2_hg38_intron_pairs_df.txt', sep='\\t', index=False, header=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## intron pairs for S2 Drosophila samples\n",
    "\n",
    "# install splice sites from K562 intron files\n",
    "S2_intronFile = open('/path/to/annotation_files/dm6_RefSeq_introns_parsed.bed')\n",
    "\n",
    "# make a dataframe of intron coordinates\n",
    "S2_introns_bedtool = dmel_introns(S2_intronFile) \n",
    "\n",
    "# import alignment files\n",
    "S2_1a_bam_file = pybedtools.BedTool('/path/to/S2_4sUchr_ONT_1a_dm6_minimap2_uniq_sort.bam')\n",
    "S2_1b_bam_file = pybedtools.BedTool('/path/to/S2_4sUchr_ONT_1b_dm6_minimap2_uniq_sort.bam')\n",
    "S2_2_bam_file = pybedtools.BedTool('/path/to/S2_4sUchr_ONT_2_dm6_minimap2_uniq_sort.bam')\n",
    "S2_3_bam_file = pybedtools.BedTool('/path/to/S2_4sUchr_ONT_3_dm6_minimap2_uniq_sort.bam')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get reads that intersect introns\n",
    "S2_1a_intersect = get_intron_intersect(S2_introns_bedtool, S2_1a_bam_file)\n",
    "S2_1b_intersect = get_intron_intersect(S2_introns_bedtool, S2_1b_bam_file)\n",
    "S2_2_intersect = get_intron_intersect(S2_introns_bedtool, S2_2_bam_file)\n",
    "S2_3_intersect = get_intron_intersect(S2_introns_bedtool, S2_3_bam_file)\n",
    "\n",
    "# get splicing information for every read that spans an intron\n",
    "S2_1a_splice_info = get_splicing_info(S2_1a_intersect,min_overlap)\n",
    "S2_1b_splice_info = get_splicing_info(S2_1b_intersect,min_overlap)\n",
    "S2_2_splice_info = get_splicing_info(S2_2_intersect,min_overlap)\n",
    "S2_3_splice_info = get_splicing_info(S2_3_intersect,min_overlap)\n",
    "\n",
    "# rerun with only introns that overlap the read without any splicing to control for read length bias\n",
    "S2_1a_splice_info = S2_1a_splice_info[S2_1a_splice_info['splice_status'] != \"UNDETERMINED\"].reset_index(drop=True)\n",
    "S2_1a_splice_info_wiOverlap = S2_1a_splice_info[S2_1a_splice_info['read_overlap']>(S2_1a_splice_info['intron_end']-S2_1a_splice_info['intron_start'])].reset_index(drop=True)\n",
    "\n",
    "S2_1b_splice_info = S2_1b_splice_info[S2_1b_splice_info['splice_status'] != \"UNDETERMINED\"].reset_index(drop=True)\n",
    "S2_1b_splice_info_wiOverlap = S2_1b_splice_info[S2_1b_splice_info['read_overlap']>(S2_1b_splice_info['intron_end']-S2_1b_splice_info['intron_start'])].reset_index(drop=True)\n",
    "\n",
    "S2_2_splice_info = S2_2_splice_info[S2_2_splice_info['splice_status'] != \"UNDETERMINED\"].reset_index(drop=True)\n",
    "S2_2_splice_info_wiOverlap = S2_2_splice_info[S2_2_splice_info['read_overlap']>(S2_2_splice_info['intron_end']-S2_2_splice_info['intron_start'])].reset_index(drop=True)\n",
    "\n",
    "S2_3_splice_info = S2_3_splice_info[S2_3_splice_info['splice_status'] != \"UNDETERMINED\"].reset_index(drop=True)\n",
    "S2_3_splice_info_wiOverlap = S2_3_splice_info[S2_3_splice_info['read_overlap']>(S2_3_splice_info['intron_end']-S2_3_splice_info['intron_start'])].reset_index(drop=True)\n",
    "\n",
    "\n",
    "# get dictionary with all intron junctions that a read spans\n",
    "S2_1a_splice_dictionary = get_read_junctions_dictionary(S2_1a_splice_info_wiOverlap)\n",
    "S2_1b_splice_dictionary = get_read_junctions_dictionary(S2_1b_splice_info_wiOverlap)\n",
    "S2_2_splice_dictionary = get_read_junctions_dictionary(S2_2_splice_info_wiOverlap)\n",
    "S2_3_splice_dictionary = get_read_junctions_dictionary(S2_3_splice_info_wiOverlap)\n",
    "\n",
    "# save dictionary with all intron junctions that a read spans\n",
    "np.save('/path/to/S2_1a_splicing_dictionary.npy', S2_1a_splice_dictionary) \n",
    "np.save('/path/to/S2_1b_splicing_dictionary.npy', S2_1b_splice_dictionary) \n",
    "np.save('/path/to/S2_2_splicing_dictionary.npy', S2_2_splice_dictionary) \n",
    "np.save('/path/to/S2_3_splicing_dictionary.npy', S2_3_splice_dictionary) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get information about intron pairs from read junctions dictionary\n",
    "S2_1a_intron_pairs_df = get_intron_pairs_df(S2_1a_splice_dictionary)\n",
    "S2_1b_intron_pairs_df = get_intron_pairs_df(S2_1b_splice_dictionary)\n",
    "S2_2_intron_pairs_df = get_intron_pairs_df(S2_2_splice_dictionary)\n",
    "S2_3_intron_pairs_df = get_intron_pairs_df(S2_3_splice_dictionary)\n",
    "\n",
    "# save splice dataframes to file (to use again later for plotting) (S2 cells)\n",
    "S2_1a_intron_pairs_df.to_csv('/path/to/S2_1a_dm6_intron_pairs_df.txt', sep='\\t', index=False, header=True)\n",
    "S2_1b_intron_pairs_df.to_csv('/path/to/S2_1b_dm6_intron_pairs_df.txt', sep='\\t', index=False, header=True)\n",
    "S2_2_intron_pairs_df.to_csv('/path/to/S2_2_dm6_intron_pairs_df.txt', sep='\\t', index=False, header=True)\n",
    "S2_3_intron_pairs_df.to_csv('/path/to/S2_3_dm6_intron_pairs_df.txt', sep='\\t', index=False, header=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Splicing dictionaries for DMSO and PlaB datasets\n",
    "K562_DMSO_1_bamFile = pybedtools.BedTool('/path/to/K562_ONT_DMSO_1_hg38_minimap2_uniq_sort.bam')\n",
    "K562_DMSO_2_bamFile = pybedtools.BedTool('/path/to/K562_ONT_DMSO_2_hg38_minimap2_uniq_sort.bam')\n",
    "K562_PlaB_1_bamFile = pybedtools.BedTool('/path/to/K562_ONT_PlaB_1_hg38_minimap2_uniq_sort.bam')\n",
    "K562_PlaB_2_bamFile = pybedtools.BedTool('/path/to/K562_ONT_PlaB_2_hg38_minimap2_uniq_sort.bam')\n",
    "\n",
    "S2_DMSO_1_bamFile = pybedtools.BedTool('/path/to/S2_ONT_DMSO_1_dm6_minimap2_uniq_sort.bam')\n",
    "S2_DMSO_2_bamFile = pybedtools.BedTool('/path/to/S2_ONT_DMSO_1_dm6_minimap2_uniq_sort.bam')\n",
    "S2_PlaB_1a_bamFile = pybedtools.BedTool('/path/to/S2_ONT_PlaB_1a_dm6_minimap2_uniq_sort.bam')\n",
    "S2_PlaB_1b_bamFile = pybedtools.BedTool('/path/to/S2_ONT_PlaB_1b_dm6_minimap2_uniq_sort.bam')\n",
    "S2_PlaB_2_bamFile = pybedtools.BedTool('/path/to/S2_ONT_PlaB_2_dm6_minimap2_uniq_sort.bam')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# K562\n",
    "# get reads that intersect introns\n",
    "K562_DMSO_1_intersect = get_intron_intersect(S2_introns_bedtool, K562_DMSO_1_bam_file)\n",
    "K562_DMSO_2_intersect = get_intron_intersect(S2_introns_bedtool, K562_DMSO_2_bam_file)\n",
    "K562_PlaB_1_intersect = get_intron_intersect(S2_introns_bedtool, K562_PlaB_1_bam_file)\n",
    "K562_PlaB_2_intersect = get_intron_intersect(S2_introns_bedtool, K562_PlaB_2_bam_file)\n",
    "\n",
    "# get splicing information for every read that spans an intron\n",
    "K562_DMSO_1_splice_info = get_splicing_info(K562_DMSO_1_intersect,min_overlap)\n",
    "K562_DMSO_2_splice_info = get_splicing_info(K562_DMSO_2_intersect,min_overlap)\n",
    "K562_PlaB_1_splice_info = get_splicing_info(K562_PlaB_1_intersect,min_overlap)\n",
    "K562_PlaB_2_splice_info = get_splicing_info(K562_PlaB_2_intersect,min_overlap)\n",
    "\n",
    "# rerun with only introns that overlap the read without any splicing to control for read length bias\n",
    "K562_DMSO_1_splice_info = K562_DMSO_1_splice_info[K562_DMSO_1_splice_info['splice_status'] != \"UNDETERMINED\"].reset_index(drop=True)\n",
    "K562_DMSO_1_splice_info_wiOverlap = K562_DMSO_1_splice_info[K562_DMSO_1_splice_info['read_overlap']>(K562_DMSO_1_splice_info['intron_end']-K562_DMSO_1_splice_info['intron_start'])].reset_index(drop=True)\n",
    "\n",
    "K562_DMSO_2_splice_info = K562_DMSO_2_splice_info[K562_DMSO_2_splice_info['splice_status'] != \"UNDETERMINED\"].reset_index(drop=True)\n",
    "K562_DMSO_2_splice_info_wiOverlap = K562_DMSO_2_splice_info[K562_DMSO_2_splice_info['read_overlap']>(K562_DMSO_2_splice_info['intron_end']-K562_DMSO_2_splice_info['intron_start'])].reset_index(drop=True)\n",
    "\n",
    "K562_PlaB_1_splice_info = K562_PlaB_1_splice_info[K562_PlaB_1_splice_info['splice_status'] != \"UNDETERMINED\"].reset_index(drop=True)\n",
    "K562_PlaB_1_splice_info_wiOverlap = K562_PlaB_1_splice_info[K562_PlaB_1_splice_info['read_overlap']>(K562_PlaB_1_splice_info['intron_end']-K562_PlaB_1_splice_info['intron_start'])].reset_index(drop=True)\n",
    "\n",
    "K562_PlaB_2_splice_info = K562_PlaB_2_splice_info[K562_PlaB_2_splice_info['splice_status'] != \"UNDETERMINED\"].reset_index(drop=True)\n",
    "K562_PlaB_2_splice_info_wiOverlap = K562_PlaB_2_splice_info[K562_PlaB_2_splice_info['read_overlap']>(K562_PlaB_2_splice_info['intron_end']-K562_PlaB_2_splice_info['intron_start'])].reset_index(drop=True)\n",
    "\n",
    "\n",
    "# get dictionary with all intron junctions that a read spans\n",
    "K562_DMSO_1_splice_dictionary = get_read_junctions_dictionary(K562_DMSO_1_splice_info_wiOverlap)\n",
    "K562_DMSO_2_splice_dictionary = get_read_junctions_dictionary(K562_DMSO_2_splice_info_wiOverlap)\n",
    "K562_PlaB_1_splice_dictionary = get_read_junctions_dictionary(K562_PlaB_1_splice_info_wiOverlap)\n",
    "K562_PlaB_2_splice_dictionary = get_read_junctions_dictionary(K562_PlaB_2_splice_info_wiOverlap)\n",
    "\n",
    "# save dictionary with all intron junctions that a read spans\n",
    "np.save('/path/to/K562_DMSO_1_splicing_dictionary.npy', K562_DMSO_1_splice_dictionary) \n",
    "np.save('/path/to/K562_DMSO_2_splicing_dictionary.npy', K562_DMSO_2_splice_dictionary) \n",
    "np.save('/path/to/K562_PlaB_1_splicing_dictionary.npy', K562_PlaB_1_splice_dictionary) \n",
    "np.save('/path/to/K562_PlaB_2_splicing_dictionary.npy', K562_PlaB_2_splice_dictionary) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# S2\n",
    "# get reads that intersect introns\n",
    "S2_DMSO_1_intersect = get_intron_intersect(S2_introns_bedtool, S2_DMSO_1_bam_file)\n",
    "S2_DMSO_2_intersect = get_intron_intersect(S2_introns_bedtool, S2_DMSO_2_bam_file)\n",
    "S2_PlaB_1a_intersect = get_intron_intersect(S2_introns_bedtool, S2_PlaB_1a_bam_file)\n",
    "S2_PlaB_1b_intersect = get_intron_intersect(S2_introns_bedtool, S2_PlaB_1b_bam_file)\n",
    "S2_PlaB_2_intersect = get_intron_intersect(S2_introns_bedtool, S2_PlaB_2_bam_file)\n",
    "\n",
    "# get splicing information for every read that spans an intron\n",
    "S2_DMSO_1_splice_info = get_splicing_info(S2_DMSO_1_intersect,min_overlap)\n",
    "S2_DMSO_2_splice_info = get_splicing_info(S2_DMSO_2_intersect,min_overlap)\n",
    "S2_PlaB_1a_splice_info = get_splicing_info(S2_PlaB_1a_intersect,min_overlap)\n",
    "S2_PlaB_1b_splice_info = get_splicing_info(S2_PlaB_1b_intersect,min_overlap)\n",
    "S2_PlaB_2_splice_info = get_splicing_info(S2_PlaB_2_intersect,min_overlap)\n",
    "\n",
    "# rerun with only introns that overlap the read without any splicing to control for read length bias\n",
    "S2_DMSO_1_splice_info = S2_DMSO_1_splice_info[S2_DMSO_1_splice_info['splice_status'] != \"UNDETERMINED\"].reset_index(drop=True)\n",
    "S2_DMSO_1_splice_info_wiOverlap = S2_DMSO_1_splice_info[S2_DMSO_1_splice_info['read_overlap']>(S2_DMSO_1_splice_info['intron_end']-S2_DMSO_1_splice_info['intron_start'])].reset_index(drop=True)\n",
    "\n",
    "S2_DMSO_2_splice_info = S2_DMSO_2_splice_info[S2_DMSO_2_splice_info['splice_status'] != \"UNDETERMINED\"].reset_index(drop=True)\n",
    "S2_DMSO_2_splice_info_wiOverlap = S2_DMSO_2_splice_info[S2_DMSO_2_splice_info['read_overlap']>(S2_DMSO_2_splice_info['intron_end']-S2_DMSO_2_splice_info['intron_start'])].reset_index(drop=True)\n",
    "\n",
    "S2_PlaB_1a_splice_info = S2_PlaB_1a_splice_info[S2_PlaB_1a_splice_info['splice_status'] != \"UNDETERMINED\"].reset_index(drop=True)\n",
    "S2_PlaB_1a_splice_info_wiOverlap = S2_PlaB_1a_splice_info[S2_PlaB_1a_splice_info['read_overlap']>(S2_PlaB_1a_splice_info['intron_end']-S2_PlaB_1a_splice_info['intron_start'])].reset_index(drop=True)\n",
    "\n",
    "S2_PlaB_1b_splice_info = S2_PlaB_1b_splice_info[S2_PlaB_1b_splice_info['splice_status'] != \"UNDETERMINED\"].reset_index(drop=True)\n",
    "S2_PlaB_1b_splice_info_wiOverlap = S2_PlaB_1b_splice_info[S2_PlaB_1b_splice_info['read_overlap']>(S2_PlaB_1b_splice_info['intron_end']-S2_PlaB_1b_splice_info['intron_start'])].reset_index(drop=True)\n",
    "\n",
    "S2_PlaB_2_splice_info = S2_PlaB_2_splice_info[S2_PlaB_2_splice_info['splice_status'] != \"UNDETERMINED\"].reset_index(drop=True)\n",
    "S2_PlaB_2_splice_info_wiOverlap = S2_PlaB_2_splice_info[S2_PlaB_2_splice_info['read_overlap']>(S2_PlaB_2_splice_info['intron_end']-S2_PlaB_2_splice_info['intron_start'])].reset_index(drop=True)\n",
    "\n",
    "\n",
    "# get dictionary with all intron junctions that a read spans\n",
    "S2_DMSO_1_splice_dictionary = get_read_junctions_dictionary(S2_DMSO_1_splice_info_wiOverlap)\n",
    "S2_DMSO_2_splice_dictionary = get_read_junctions_dictionary(S2_DMSO_2_splice_info_wiOverlap)\n",
    "S2_PlaB_1a_splice_dictionary = get_read_junctions_dictionary(S2_PlaB_1a_splice_info_wiOverlap)\n",
    "S2_PlaB_1b_splice_dictionary = get_read_junctions_dictionary(S2_PlaB_1b_splice_info_wiOverlap)\n",
    "S2_PlaB_2_splice_dictionary = get_read_junctions_dictionary(S2_PlaB_2_splice_info_wiOverlap)\n",
    "\n",
    "# save dictionary with all intron junctions that a read spans\n",
    "np.save('/path/to/S2_DMSO_1_splicing_dictionary.npy', S2_DMSO_1_splice_dictionary) \n",
    "np.save('/path/to/S2_DMSO_2_splicing_dictionary.npy', S2_DMSO_2_splice_dictionary) \n",
    "np.save('/path/to/S2_PlaB_1a_splicing_dictionary.npy', S2_PlaB_1a_splice_dictionary)\n",
    "np.save('/path/to/S2_PlaB_1b_splicing_dictionary.npy', S2_PlaB_1b_splice_dictionary) \n",
    "np.save('/path/to/S2_PlaB_2_splicing_dictionary.npy', S2_PlaB_2_splice_dictionary) "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
