{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "\n",
    "Date : August 1, 2019\n",
    "\n",
    "Author : Heather Landry Drexler\n",
    "\n",
    "This script will develop datasets for making plots for percent spliced\n",
    "terminal introns at termination sites from nanoCOP data.\n",
    "\n",
    "These files serve as input for the scripts for Figures 2, 3 and S5.\n",
    "                                       \n",
    "                                            \n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import re\n",
    "import math\n",
    "import pybedtools\n",
    "from pybedtools import BedTool\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_last_polyA_site(last_intron_df):\n",
    "    \n",
    "    df = last_intron_df.copy()\n",
    "    df['gene'] = df.apply(lambda row: row.gene_polyA.split('_')[0]+'_'+row.gene_polyA.split('_')[1] ,axis=1)\n",
    "    df['polyA'] = df.apply(lambda row: row.gene_polyA.split('_')[2] ,axis=1)\n",
    "    df = df.drop(['gene_polyA'], axis = 1)\n",
    "\n",
    "    polyA_sites = {}\n",
    "\n",
    "    for i in range(len(df)):\n",
    "\n",
    "        chrom = df['chrom'].iloc[i]\n",
    "        start = df['start'].iloc[i]\n",
    "        end = df['end'].iloc[i]\n",
    "        strand = df['strand'].iloc[i]\n",
    "        polyA = int(df['polyA'].iloc[i])\n",
    "\n",
    "        coord = chrom+'_'+str(start)+'_'+str(end)+'_'+str(strand)\n",
    "\n",
    "        if coord not in polyA_sites.keys():\n",
    "\n",
    "            polyA_sites[coord] = polyA\n",
    "\n",
    "\n",
    "        if coord in polyA_sites.keys():\n",
    "\n",
    "            if strand==\"+\" and polyA>polyA_sites[coord]:\n",
    "                polyA_sites[coord] = polyA\n",
    "\n",
    "            if strand==\"-\" and polyA<polyA_sites[coord]:\n",
    "                polyA_sites[coord] = polyA\n",
    "\n",
    "    last_polyA_site = []\n",
    "\n",
    "    for key in polyA_sites: \n",
    "\n",
    "        chrom = key.split('_')[0]\n",
    "        start = key.split('_')[1]\n",
    "        end = key.split('_')[2]\n",
    "        strand = key.split('_')[3]\n",
    "        polyA = polyA_sites[key]\n",
    "\n",
    "        last_polyA_site.append([chrom, start, end, strand, polyA])\n",
    "\n",
    "    last_polyA_site_df = pd.DataFrame(last_polyA_site)\n",
    "    last_polyA_site_df.columns = ['chrom','start','end','strand','polyA']\n",
    "\n",
    "    return last_polyA_site_df\n",
    "\n",
    "\n",
    "# function to get a bedtool file with splice site info from hg38 intron coordinate bed file\n",
    "def human_spliceSites(last_intron_df):\n",
    "    # make a set for all 3'SS coordinates\n",
    "    introns = []\n",
    "\n",
    "    # loop through a file with intron coordinates\n",
    "    # record features about the introns\n",
    "    for i in range(0,len(last_intron_df)):\n",
    "        chrom = str(last_intron_df['chrom'].iloc[i])  # chromosome\n",
    "        start = int(last_intron_df['start'].iloc[i])        # start coordinate of intron (last base of exon)\n",
    "        end = int(last_intron_df['end'].iloc[i])            # end coordinate of intron (last base of intron)\n",
    "        polyA = int(last_intron_df['polyA'].iloc[i])        # gene name and polyA site for that gene (multiples are split with \"_\")\n",
    "        strand = last_intron_df['strand'].iloc[i]           # strand of gene with intron\n",
    "\n",
    "        # get 3' SS positions for introns plus one base\n",
    "        if strand=='+':\n",
    "            pos1 = int(end)\n",
    "            pos2 = int(end)+1\n",
    "            pos5prime = int(start)+1\n",
    "        if strand=='-':\n",
    "            pos1 = int(start)\n",
    "            pos2 = int(start)+1\n",
    "            pos5prime = int(end)\n",
    "\n",
    "        # make a key that will represent intron coordinates\n",
    "        introns.append([str(chrom),str(pos1),str(pos2),str(polyA),str(pos5prime),str(strand)])\n",
    "\n",
    "    spliceSites = BedTool(introns)\n",
    "    return spliceSites\n",
    "\n",
    "\n",
    "# function to get a bedtool file with splice site info from dmel6 intron coordinate bed file\n",
    "def drosophila_spliceSites(last_intron_df):\n",
    "    # make a set for all 3'SS coordinates\n",
    "    introns = []\n",
    "\n",
    "    # loop through a file with intron coordinates\n",
    "    # record features about the introns\n",
    "    for i in range(0,len(last_intron_df)):\n",
    "        chrom = 'chr'+str(last_intron_df['chrom'].iloc[i])  # chromosome\n",
    "        start = int(last_intron_df['start'].iloc[i])        # start coordinate of intron (last base of exon)\n",
    "        end = int(last_intron_df['end'].iloc[i])            # end coordinate of intron (last base of intron)\n",
    "        polyA = int(last_intron_df['polyA'].iloc[i])        # gene name and polyA site for that gene (multiples are split with \"_\")\n",
    "        strand = last_intron_df['strand'].iloc[i]           # strand of gene with intron\n",
    "\n",
    "        # get 3' SS positions for introns plus one base\n",
    "        if strand=='+':\n",
    "            pos1 = int(end)\n",
    "            pos2 = int(end)+1\n",
    "            pos5prime = int(start)+1\n",
    "        if strand=='-':\n",
    "            pos1 = int(start)\n",
    "            pos2 = int(start)+1\n",
    "            pos5prime = int(end)\n",
    "\n",
    "        # make a key that will represent intron coordinates\n",
    "        introns.append([str(chrom),str(pos1),str(pos2),str(polyA),str(pos5prime),str(strand)])\n",
    "\n",
    "    spliceSites = BedTool(introns)\n",
    "    return spliceSites\n",
    "\n",
    "\n",
    "# function to create a dataframe with reads that span 3'SS positions\n",
    "def get_spliceSite_df(spliceSites, bamFile):\n",
    "    # get reads that span 3' splice sites and convert to a dataframe\n",
    "    bedFile = bamFile.bam_to_bed(cigar=True, tag='NM') # convert bam file to bed file, keep cigar string and NM (edit distance) tag\n",
    "    intersect = bedFile.intersect(spliceSites, wo=True, s=True) # intersect reads from bam file with 3' splice site coordinates, ensure strandedness\n",
    "    df = intersect.to_dataframe(names=['chrom', 'start_aln', 'end_aln', 'name_aln', 'qual_aln', \\\n",
    "                                           'strand_aln', 'cigar_aln', 'chr_3SS', 'start_3SS', \\\n",
    "                                           'end_3SS', 'polyA', 'pos_5SS', 'strand_gene', 'count'], \\\n",
    "                               dtype={\"chrom\": str, \"start_aln\": int, \"end_aln\": int, \\\n",
    "                                     \"name_aln\": str, \"qual_aln\": int, \"strand_aln\": str, \\\n",
    "                                     \"cigar_aln\": str, \"chr_3SS\": str, \"start_3SS\": int, \\\n",
    "                                     \"end_3SS\": int, \"polyA\": int, \\\n",
    "                                     \"pos_5SS\": int,\"strand_gene\": str, \"count\": int}) # convert to a dataframe\n",
    "    return df\n",
    "\n",
    "\n",
    "# function to get a splicing calls from intersect dataframe\n",
    "def get_MinION_spliceCalls(df, min_overlap):\n",
    "    \n",
    "    # prepare a list for splice calls\n",
    "    spliceCalls = []\n",
    "\n",
    "    # set variables for parsing the cigar string\n",
    "    pattern = re.compile('([MIDNSHPX=])')\n",
    "    Consumes_Query = [\"M\", \"I\", \"S\", \"=\", \"X\"]\n",
    "    Consumes_Reference = [\"M\", \"D\", \"N\", \"=\", \"X\"]    \n",
    "\n",
    "    for i in range(0,df.shape[0]):\n",
    "        if df.strand_gene[i] == \"-\":\n",
    "            align_3p_end = df.start_aln[i] # record 3' end of read for - strand genes\n",
    "            align_5p_end = df.end_aln[i] # record 5' end of read for - strand genes\n",
    "            pos_3SS = df.end_3SS[i] # record 3'SS position for - strand genes\n",
    "            pos_5SS = df.pos_5SS[i] # record 5'SS position for - strand genes\n",
    "            intron_start = pos_3SS # get position for the start of intron coordinate on negative strand\n",
    "            intron_end = pos_5SS # get position for the end of intron coordinate on negative strand\n",
    "\n",
    "        if df.strand_gene[i] == \"+\":\n",
    "            align_3p_end = df.end_aln[i] # record 3' end of read for + strand genes\n",
    "            align_5p_end = df.start_aln[i] # record 5' end of read for + strand genes\n",
    "            pos_3SS = df.start_3SS[i] # record 3'SS position for + strand genes \n",
    "            pos_5SS = df.pos_5SS[i] # record 5'SS position for - strand genes \n",
    "            intron_start = pos_5SS # get position for the start of intron coordinate on positive strand\n",
    "            intron_end = pos_3SS # get position for the end of intron coordinate on positive strand\n",
    "            \n",
    "        # calculate distance between 3'SS and 3'end of read \n",
    "        dist = abs(align_3p_end - pos_3SS) #*** double check this!!!\n",
    "\n",
    "        # parse cigar string into a list of tuples for easy parsing\n",
    "        Sep_Values = pattern.split(df.cigar_aln[i])[:-1]\n",
    "        CigarPairs = list((Sep_Values[n:n+2] for n in range(0, len(Sep_Values), 2)))  \n",
    "\n",
    "        # get the 3' softclip length\n",
    "        if df.strand_aln[i]==\"+\":        \n",
    "            last=len(CigarPairs)\n",
    "            if(CigarPairs[last-1][1]=='S'):\n",
    "                clip=CigarPairs[last-1][0]\n",
    "            elif(CigarPairs[last-1][1]=='H'):\n",
    "                clip=CigarPairs[last-1][0]\n",
    "            elif(CigarPairs[last-1][1]!='S' or CigarPairs[last-1][1]!='H'):\n",
    "                clip=0\n",
    "\n",
    "        if df.strand_aln[i]==\"-\":\n",
    "            if(CigarPairs[0][1]=='S'):\n",
    "                clip=CigarPairs[0][0]\n",
    "            elif(CigarPairs[0][1]=='H'):\n",
    "                clip=CigarPairs[0][0]\n",
    "            elif(CigarPairs[0][1]!='S' or CigarPairs[0][1]!='H'):\n",
    "                clip=0\n",
    "\n",
    "        # set up variables for measuring the length of cigar string operators\n",
    "        CigarOp_counts = {'M': 0, 'I': 0, 'D': 0, 'N': 0, 'S': 0, 'H': 0, 'P': 0, '=': 0, 'X': 0}\n",
    "        start_counts = {'M': 0, 'I': 0, 'D': 0, 'N': 0, 'S': 0, 'H': 0, 'P': 0, '=': 0, 'X': 0}\n",
    "        end_counts = {'M': 0, 'I': 0, 'D': 0, 'N': 0, 'S': 0, 'H': 0, 'P': 0, '=': 0, 'X': 0}\n",
    "        intron_counts = {'M': 0, 'I': 0, 'D': 0, 'N': 0, 'S': 0, 'H': 0, 'P': 0, '=': 0, 'X': 0}\n",
    "        currentloc = int(df.start_aln[i]) \n",
    "\n",
    "        # go through list of cigar strings and grab splicing information\n",
    "        for cigar_Entry in CigarPairs:\n",
    "\n",
    "            op_Length = int(cigar_Entry[0]) # get length of cigar operator\n",
    "            cigarOp = cigar_Entry[1] # get type of cigar operator  \n",
    "            CigarOp_counts[cigarOp] += op_Length # add the cigar operator length to the counts dictionary\n",
    "            cigarOp_start=currentloc # get the starting coordinate of the cigar operator\n",
    "\n",
    "            if (cigarOp in Consumes_Reference):\n",
    "                currentloc=currentloc+op_Length # add the cigar operator length to the current location coordinate \n",
    "\n",
    "            cigarOp_end=currentloc # get the ending coordinate of the cigar operator\n",
    "\n",
    "            # gather information if the portion of the cigar string spans the designated 5' splice site\n",
    "            if (cigarOp_start<intron_start-min_overlap and cigarOp_end>=intron_start-min_overlap):\n",
    "                if (cigarOp_end>=intron_start+min_overlap):\n",
    "                    count=min_overlap*2\n",
    "                else:\n",
    "                    count=cigarOp_end-(intron_start-min_overlap)+1\n",
    "                start_counts[cigarOp] += count # add the cigar operator length to the counts dictionary\n",
    "\n",
    "            elif (cigarOp_start>=intron_start-min_overlap and cigarOp_end<intron_start+min_overlap):\n",
    "                count=op_Length\n",
    "                start_counts[cigarOp] += count # add the cigar operator length to the counts dictionary       \n",
    "\n",
    "            elif (cigarOp_start<intron_start+min_overlap and cigarOp_end>=intron_start+min_overlap):\n",
    "                if (cigarOp_start<=intron_start-min_overlap):\n",
    "                    count=min_overlap*2\n",
    "                else:\n",
    "                    count=(intron_start+min_overlap)-cigarOp_start-1\n",
    "                start_counts[cigarOp] += count # add the cigar operator length to the counts dictionary \n",
    "\n",
    "            # gather information if the portion of the cigar string is within the intron\n",
    "            if (cigarOp_start<intron_start and cigarOp_end>=intron_start):\n",
    "                if (cigarOp_end>=intron_end):\n",
    "                    count=intron_end-intron_start\n",
    "                else:\n",
    "                    count=cigarOp_end-intron_start\n",
    "                intron_counts[cigarOp] += count # add the cigar operator length to the counts dictionary\n",
    "\n",
    "            elif (cigarOp_start>=intron_start and cigarOp_end<intron_end):\n",
    "                count=op_Length\n",
    "                intron_counts[cigarOp] += count # add the cigar operator length to the counts dictionary\n",
    "\n",
    "            elif (cigarOp_start<intron_end and cigarOp_end>=intron_end):\n",
    "                if (cigarOp_start<=intron_start):\n",
    "                    count=intron_end-intron_start\n",
    "                else:\n",
    "                    count=intron_end-cigarOp_start\n",
    "                intron_counts[cigarOp] += count # add the cigar operator length to the counts dictionary \n",
    "\n",
    "            # gather information if the portion of the cigar string spans the designated 3' splice site\n",
    "            if (cigarOp_start<intron_end-min_overlap and cigarOp_end>=intron_end-min_overlap):\n",
    "                if (cigarOp_end>=intron_end+min_overlap):\n",
    "                    count=min_overlap*2\n",
    "                else:\n",
    "                    count=cigarOp_end-(intron_end-min_overlap)\n",
    "                end_counts[cigarOp] += count # add the cigar operator length to the counts dictionary\n",
    "\n",
    "            elif (cigarOp_start>=intron_end-min_overlap and cigarOp_end<intron_end+min_overlap):\n",
    "                count=op_Length\n",
    "                end_counts[cigarOp] += count # add the cigar operator length to the counts dictionary\n",
    "\n",
    "            elif (cigarOp_start<intron_end+min_overlap and cigarOp_end>=intron_end+min_overlap):\n",
    "                if (cigarOp_start<=intron_end-min_overlap):\n",
    "                    count=min_overlap*2\n",
    "                else:\n",
    "                    count=(intron_end+min_overlap)-cigarOp_start\n",
    "                end_counts[cigarOp] += count # add the cigar operator length to the counts dictionary \n",
    "\n",
    "\n",
    "        # assign strandedness to determine counts around 5'SS and 3'SS\n",
    "        if(df.strand_gene[i]=='+'):\n",
    "            around5SS_counts = start_counts\n",
    "            around3SS_counts = end_counts\n",
    "\n",
    "        elif(df.strand_gene[i]==\"-\"):\n",
    "            around5SS_counts = end_counts\n",
    "            around3SS_counts = start_counts\n",
    "\n",
    "        # annotate splicing status based on CIGAR string information around splice sites\n",
    "        if(around3SS_counts['N']==0 and around3SS_counts['M']>min_overlap/2):\n",
    "            splice='NO'\n",
    "        elif(around3SS_counts['N']>0 and around3SS_counts['N']<min_overlap*2):\n",
    "            if(around5SS_counts['N']>0 and around5SS_counts['N']<min_overlap*2):\n",
    "                splice='YES'\n",
    "            else:\n",
    "                splice='UNDETERMINED'\n",
    "        else:\n",
    "            splice='UNDETERMINED'\n",
    "\n",
    "        # annotate splicing status based on CIGAR string information within the intron \n",
    "        if (splice == 'YES'):\n",
    "            if (float(intron_end-intron_start) > 0.0):\n",
    "                ratio = float(intron_counts['N'])/float(intron_end-intron_start)\n",
    "                difference = abs(intron_counts['N']-(intron_end-intron_start))\n",
    "                # if read is spliced, between 90-100% of the intron has to be spliced \n",
    "                # and no more than 100 nucleotides within the intron can be matching the intron sequence\n",
    "                if( ratio < 0.9 or ratio > 1.1 or difference > 100):\n",
    "                    splice='UNDETERMINED'\n",
    "            if (float(intron_end-intron_start) == 0.0):\n",
    "                splice='UNDETERMINED'\n",
    "\n",
    "        if (splice == 'NO'):\n",
    "            if (float(intron_end-intron_start) > 0.0):\n",
    "                intronLength = intron_end-intron_start\n",
    "                difference = abs(intron_counts['M']+intron_counts['D']+intron_counts['S']-intronLength)\n",
    "                ratio = float(intron_counts['M'])/(float(intron_counts['M'])+float(intron_counts['N'])+float(intron_counts['D'])+1)\n",
    "                # if read is unspliced, at least 70% of the read has to match (CIGAR=M) the intron sequence\n",
    "                # and at least 10 nucleotides must match (CIGAR=M) within the intron sequence\n",
    "                if(intron_counts['M'] < 10 or ratio < 0.7):\n",
    "                    splice='UNDETERMINED'\n",
    "            if (float(intron_end-intron_start) == 0.0):\n",
    "                splice='UNDETERMINED'\n",
    "\n",
    "        # get information on match percentage / error length \n",
    "        # this will be used as a quality control cutoff if necessary\n",
    "        read_length = CigarOp_counts['M']+CigarOp_counts['I']+CigarOp_counts['S']+CigarOp_counts['=']+CigarOp_counts['X']\n",
    "        error_rate = float(df.qual_aln[i])/float(read_length)\n",
    "\n",
    "        spliceCalls.append([df.name_aln[i],df.chrom[i],str(intron_start),str(intron_end),str(align_5p_end),str(align_3p_end),str(read_length),df.strand_gene[i],str(error_rate),str(clip),str(dist),splice,df.polyA[i]])\n",
    "    \n",
    "    spliceCalls_df = pd.DataFrame(spliceCalls)\n",
    "    spliceCalls_df.columns = [\"read_name\",\"chrom\",\"intron_start\",\"intron_end\",\"read_start\",\"read_end\",\"read_length\",\"strand\",\"error_rate\",\"end_clippling\",\"dist_from_3SS\",\"splice_status\",\"polyA\"]\n",
    "   \n",
    "    return spliceCalls_df\n",
    "\n",
    "\n",
    "# function to line up splicing calls with distance from polyA site information\n",
    "def get_termination_splicingCalls(splicingCalls, overlap):\n",
    "    termination = []\n",
    "    for i in range(0,splicingCalls.shape[0]):\n",
    "        read_name = splicingCalls['read_name'].iloc[i]\n",
    "        chrom = splicingCalls['chrom'].iloc[i]\n",
    "        intron_start = int(splicingCalls['intron_start'].iloc[i])\n",
    "        intron_end = int(splicingCalls['intron_end'].iloc[i])\n",
    "        strand = splicingCalls['strand'].iloc[i]\n",
    "        read_length = int(splicingCalls['read_length'].iloc[i])\n",
    "        intron_length = int(splicingCalls['intron_end'].iloc[i])-int(splicingCalls['intron_start'].iloc[i])\n",
    "        dist_from_3SS = int(splicingCalls['dist_from_3SS'].iloc[i])\n",
    "        splice_status = str(splicingCalls['splice_status'].iloc[i])\n",
    "\n",
    "        if (read_length >= (dist_from_3SS + overlap)) and (splice_status != 'UNDETERMINED'):\n",
    "        \n",
    "            if (splicingCalls['strand'].iloc[i] == '+'):\n",
    "                read_end = int(splicingCalls['read_end'].iloc[i])\n",
    "                polyA = int(splicingCalls['polyA'].iloc[i])\n",
    "                dist_from_polyA = read_end-polyA\n",
    "                termination.append([read_name, intron_length, dist_from_polyA, splice_status])\n",
    "                #termination.append([read_name, chrom, intron_start, intron_end, strand, name, polyA, dist_from_polyA, splice_status])\n",
    "\n",
    "            if (splicingCalls['strand'].iloc[i] == '-'):\n",
    "                read_end = int(splicingCalls['read_end'].iloc[i])\n",
    "                polyA = int(splicingCalls['polyA'].iloc[i])\n",
    "                dist_from_polyA = polyA-read_end\n",
    "                termination.append([read_name, intron_length, dist_from_polyA, splice_status])\n",
    "                #termination.append([read_name, chrom, intron_start, intron_end, strand, name, polyA, dist_from_polyA, splice_status])\n",
    "                                        \n",
    "    termination_df = pd.DataFrame(termination)\n",
    "    termination_df.columns = [\"read_name\",\"intron_length\",\"polyA_dist\",\"splice_status\"]\n",
    "    #termination_df.columns = [\"read_name\",\"chrom\",\"intron_start\",\"intron_end\",\"strand\",\"gene_name\",\"polyA_pos\",\"polyA_dist\",\"splice_status\"]                     \n",
    "    termination_df = termination_df.drop_duplicates(subset=['read_name','splice_status']).reset_index(drop=True)  \n",
    "    \n",
    "    return termination_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "### get splicing information at termination sites for all K562 samples\n",
    "\n",
    "# read file with last intron coordinates for polyA analysis\n",
    "hg38_last_intron_df = pd.read_csv(\"/path/to/annotation_files/hg38_last_introns_polyA.txt\",sep=\"\\t\")\n",
    "\n",
    "# get coordinates with only the last polyA site\n",
    "hg38_last_intron_last_polyA_df = get_last_polyA_site(hg38_last_intron_df)\n",
    "\n",
    "# make a bedtool with 3' splice site coordinates from the bedfile with last introns\n",
    "hg38_spliceSites = human_spliceSites(hg38_last_intron_last_polyA_df)\n",
    "\n",
    "# set variables\n",
    "min_overlap = 25\n",
    "termination_overlap = 100\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# prepare dataframes for terminal intron splicing plots\n",
    "\n",
    "# K562_1\n",
    "bamFile = pybedtools.BedTool('/path/to/K562_4sUchr_ONT_1_hg38_minimap2_uniq_sort.bam')\n",
    "spliceSites_df = get_spliceSite_df(hg38_spliceSites, bamFile)\n",
    "splicingCalls = get_MinION_spliceCalls(spliceSites_df,min_overlap)\n",
    "K562_1_termination_df = get_termination_splicingCalls(splicingCalls, termination_overlap)\n",
    "\n",
    "# K562_2\n",
    "bamFile = pybedtools.BedTool('/path/to/K562_4sUchr_ONT_2_hg38_minimap2_uniq_sort.bam')\n",
    "spliceSites_df = get_spliceSite_df(hg38_spliceSites, bamFile)\n",
    "splicingCalls = get_MinION_spliceCalls(spliceSites_df,min_overlap)\n",
    "K562_2_termination_df = get_termination_splicingCalls(splicingCalls, termination_overlap)\n",
    "\n",
    "# K562_3\n",
    "bamFile = pybedtools.BedTool('/path/to/K562_4sUchr_ONT_3_hg38_minimap2_uniq_sort.bam')\n",
    "spliceSites_df = get_spliceSite_df(hg38_spliceSites, bamFile)\n",
    "splicingCalls = get_MinION_spliceCalls(spliceSites_df,min_overlap)\n",
    "K562_3_termination_df = get_termination_splicingCalls(splicingCalls, termination_overlap)\n",
    "\n",
    "# K562_4\n",
    "bamFile = pybedtools.BedTool('/path/to/K562_4sUchr_ONT_4_hg38_minimap2_uniq_sort.bam')\n",
    "spliceSites_df = get_spliceSite_df(hg38_spliceSites, bamFile)\n",
    "splicingCalls = get_MinION_spliceCalls(spliceSites_df,min_overlap)\n",
    "K562_4_termination_df = get_termination_splicingCalls(splicingCalls, termination_overlap)\n",
    "\n",
    "# K562_5a\n",
    "bamFile = pybedtools.BedTool('/path/to/K562_4sUchr_ONT_5a_hg38_minimap2_uniq_sort.bam')\n",
    "spliceSites_df = get_spliceSite_df(hg38_spliceSites, bamFile)\n",
    "splicingCalls = get_MinION_spliceCalls(spliceSites_df,min_overlap)\n",
    "K562_5a_termination_df = get_termination_splicingCalls(splicingCalls, termination_overlap)\n",
    "\n",
    "# K562_5b\n",
    "bamFile = pybedtools.BedTool('/path/to/K562_4sUchr_ONT_5b_hg38_minimap2_uniq_sort.bam')\n",
    "spliceSites_df = get_spliceSite_df(hg38_spliceSites, bamFile)\n",
    "splicingCalls = get_MinION_spliceCalls(spliceSites_df,min_overlap)\n",
    "K562_5b_termination_df = get_termination_splicingCalls(splicingCalls, termination_overlap)\n",
    "\n",
    "# K562_DMSO_1\n",
    "bamFile = pybedtools.BedTool('/path/to/K562_ONT_DMSO_1_hg38_minimap2_uniq_sort.bam')\n",
    "spliceSites_df = get_spliceSite_df(hg38_spliceSites, bamFile)\n",
    "splicingCalls = get_MinION_spliceCalls(spliceSites_df,min_overlap)\n",
    "K562_DMSO_1_termination_df = get_termination_splicingCalls(splicingCalls, termination_overlap)\n",
    "\n",
    "# K562_DMSO_2\n",
    "bamFile = pybedtools.BedTool('/path/to/K562_ONT_DMSO_2_hg38_minimap2_uniq_sort.bam')\n",
    "spliceSites_df = get_spliceSite_df(hg38_spliceSites, bamFile)\n",
    "splicingCalls = get_MinION_spliceCalls(spliceSites_df,min_overlap)\n",
    "K562_DMSO_2_termination_df = get_termination_splicingCalls(splicingCalls, termination_overlap)\n",
    "\n",
    "# K562_PlaB_1\n",
    "bamFile = pybedtools.BedTool('/path/to/K562_ONT_PlaB_1_hg38_minimap2_uniq_sort.bam')\n",
    "spliceSites_df = get_spliceSite_df(hg38_spliceSites, bamFile)\n",
    "splicingCalls = get_MinION_spliceCalls(spliceSites_df,min_overlap)\n",
    "K562_PlaB_1_termination_df = get_termination_splicingCalls(splicingCalls, termination_overlap)\n",
    "\n",
    "# K562_PlaB_2\n",
    "bamFile = pybedtools.BedTool('/path/to/K562_ONT_PlaB_2_hg38_minimap2_uniq_sort.bam')\n",
    "spliceSites_df = get_spliceSite_df(hg38_spliceSites, bamFile)\n",
    "splicingCalls = get_MinION_spliceCalls(spliceSites_df,min_overlap)\n",
    "K562_PlaB_2_termination_df = get_termination_splicingCalls(splicingCalls, termination_overlap)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# save splice dataframes to file (to use again later for plotting) (K562 cells)\n",
    "K562_1_termination_df.to_csv('/path/to/K562_1_hg38_termination_df.txt', sep='\\t', index=False, header=True)\n",
    "K562_2_termination_df.to_csv('/path/to/K562_2_hg38_termination_df.txt', sep='\\t', index=False, header=True)\n",
    "K562_3_termination_df.to_csv('/path/to/K562_3_hg38_termination_df.txt', sep='\\t', index=False, header=True)\n",
    "K562_4_termination_df.to_csv('/path/to/K562_4_hg38_termination_df.txt', sep='\\t', index=False, header=True)\n",
    "K562_5a_termination_df.to_csv('/path/to/K562_5a_hg38_termination_df.txt', sep='\\t', index=False, header=True)\n",
    "K562_5b_termination_df.to_csv('/path/to/K562_5b_hg38_termination_df.txt', sep='\\t', index=False, header=True)\n",
    "K562_DMSO_1_termination_df.to_csv('/path/to/K562_DMSO_1_hg38_termination_df.txt', sep='\\t', index=False, header=True)\n",
    "K562_DMSO_2_termination_df.to_csv('/path/to/K562_DMSO_2_hg38_termination_df.txt', sep='\\t', index=False, header=True)\n",
    "K562_PlaB_1_termination_df.to_csv('/path/to/K562_PlaB_1_hg38_termination_df.txt', sep='\\t', index=False, header=True)\n",
    "K562_PlaB_2_termination_df.to_csv('/path/to/K562_PlaB_2_hg38_termination_df.txt', sep='\\t', index=False, header=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### get splicing information at termination sites for all S2 samples\n",
    "\n",
    "# read file with last intron coordinates for polyA analysis\n",
    "dmel6_last_intron_df = pd.read_csv(\"/path/to/annotation_files/dmel6_last_introns_polyA.txt\",sep=\"\\t\")\n",
    "\n",
    "# get coordinates with only the last polyA site\n",
    "dmel6_last_intron_last_polyA_df = get_last_polyA_site(dmel6_last_intron_df)\n",
    "\n",
    "# make a bedtool with 3' splice site coordinates from the bedfile with last introns\n",
    "dmel6_spliceSites = drosophila_spliceSites(dmel6_last_intron_last_polyA_df)\n",
    "\n",
    "\n",
    "# set variables\n",
    "min_overlap = 25\n",
    "termination_overlap = 100\n",
    "\n",
    "# S2_1a\n",
    "bamFile = pybedtools.BedTool('/path/to/S2_4sUchr_ONT_1a_dm6_minimap2_uniq_sort.bam')\n",
    "spliceSites_df = get_spliceSite_df(dmel6_spliceSites, bamFile)\n",
    "splicingCalls = get_MinION_spliceCalls(spliceSites_df,min_overlap)\n",
    "S2_1a_termination_df = get_termination_splicingCalls(splicingCalls, termination_overlap)\n",
    "\n",
    "# S2_1b\n",
    "bamFile = pybedtools.BedTool('/path/to/S2_4sUchr_ONT_1b_dm6_minimap2_uniq_sort.bam')\n",
    "spliceSites_df = get_spliceSite_df(dmel6_spliceSites, bamFile)\n",
    "splicingCalls = get_MinION_spliceCalls(spliceSites_df,min_overlap)\n",
    "S2_1b_termination_df = get_termination_splicingCalls(splicingCalls, termination_overlap)\n",
    "\n",
    "# S2_3\n",
    "bamFile = pybedtools.BedTool('/path/to/S2_4sUchr_ONT_2_dm6_minimap2_uniq_sort.bam')\n",
    "spliceSites_df = get_spliceSite_df(dmel6_spliceSites, bamFile)\n",
    "splicingCalls = get_MinION_spliceCalls(spliceSites_df,min_overlap)\n",
    "S2_3_termination_df = get_termination_splicingCalls(splicingCalls, termination_overlap)\n",
    "\n",
    "# S2_3\n",
    "bamFile = pybedtools.BedTool('/path/to/S2_4sUchr_ONT_3_dm6_minimap2_uniq_sort.bam')\n",
    "spliceSites_df = get_spliceSite_df(dmel6_spliceSites, bamFile)\n",
    "splicingCalls = get_MinION_spliceCalls(spliceSites_df,min_overlap)\n",
    "S2_3_termination_df = get_termination_splicingCalls(splicingCalls, termination_overlap)\n",
    "\n",
    "# merge technical replicates S2_1a + S2_1b\n",
    "S2_1_termination_df = pd.concat([S2_1a_termination_df,S2_1b_termination_df])\n",
    "\n",
    "\n",
    "# save splice dataframes to file (to use again later for plotting) (K562 cells)\n",
    "S2_1a_termination_df.to_csv('/path/to/S2_1a_dm6_termination_df.txt', sep='\\t', index=False, header=True)\n",
    "S2_1b_termination_df.to_csv('/path/to/S2_1b_dm6_termination_df.txt', sep='\\t', index=False, header=True)\n",
    "S2_3_termination_df.to_csv('/path/to/S2_3_dm6_termination_df.txt', sep='\\t', index=False, header=True)\n",
    "S2_3_termination_df.to_csv('/path/to/S2_3_dm6_termination_df.txt', sep='\\t', index=False, header=True)\n",
    "S2_1_termination_df.to_csv('/path/to/S2_1_dm6_termination_df.txt', sep='\\t', index=False, header=True)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
